app/services/hashtag_autocomplete_service.rb

```diff
@@ -9,43 +9,54 @@ def self.search_conditions
   end
 
   attr_reader :guardian
-  cattr_reader :data_sources, :contexts
 
-  def self.register_data_source(type, klass)
-    @@data_sources[type] = klass
+  def self.data_sources
+    # Category and Tag data sources are in core and always should be
+    # included for searches and lookups.
+    [CategoryHashtagDataSource, TagHashtagDataSource] |
+      DiscoursePluginRegistry.hashtag_autocomplete_data_sources
   end
 
-  def self.clear_registered
-    @@data_sources = {}
-    @@contexts = {}
+  def self.contextual_type_priorities
+    # Category and Tag type priorities for the composer are default and
+    # always are included.
+    [
+      { type: "category", context: "topic-composer", priority: 100 },
```

Comment:
Given how fleshed out this data is, perhaps we could promote it to a struct. It would give us some benefits in better error messages and access to e.g. `map(&:type)` syntax. The drawback is we need to convert the data from the plugin registry here.